src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -0,0 +1,259 @@
+
+		public UnconditionalSuppressMessageAttributeState (AssemblyDefinition assembly)
+		{
+			_assembly = assembly;
+			_localSuppressionsByMdToken = new Dictionary<MetadataToken, ImmutableDictionary<string, SuppressMessageInfo>> ();
+		}
+
+		public void AddLocalSuppression (CustomAttribute ca, MetadataToken mdToken)
+		{
+			SuppressMessageInfo info;
+			if (!TryDecodeSuppressMessageAttributeData (ca, out info)) {
+				return;
+			}
+
+			if (_localSuppressionsByMdToken.ContainsKey (mdToken)) {
+				_localSuppressionsByMdToken[mdToken].Add (info.Id, info);
+			} else {
+				_localSuppressionsByMdToken[mdToken] = new Dictionary<string, SuppressMessageInfo> { { info.Id, info } }.ToImmutableDictionary ();
+			}
+		}
+
+		public bool IsSuppressed (string id, MessageOrigin warningOrigin, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			if (HasGlobalSuppressions && IsGloballySuppressed (id, out info)) {
+				return true;
+			}
+
+			if (HasSuppressions && warningOrigin.MdTokenProvider != null) {
+				IMetadataTokenProvider mdTokenProvider = warningOrigin.MdTokenProvider;
+				while (mdTokenProvider != null) {
+					if (IsLocallySuppressed (id, mdTokenProvider.MetadataToken, out info))
+						return true;
+
+					mdTokenProvider = (mdTokenProvider as IMemberDefinition).DeclaringType;
+				}
+			}
+
+			return false;
+		}
+
+		private bool TryDecodeSuppressMessageAttributeData (CustomAttribute attribute, out SuppressMessageInfo info)
+		{
+			info = default;
+
+			// We need at least the Category and Id to decode the warning to suppress.
+			// The only UnconditionalSuppressMessageAttribute constructor requires those two parameters.
+			if (attribute.ConstructorArguments.Count < 2) {
+				return false;
+			}
+
+			// Ignore the category parameter because it does not identify the warning
+			// and category information can be obtained from warnings themselves.
+			info.Id = attribute.ConstructorArguments[1].Value as string;
+			if (info.Id == null) {
+				return false;
+			}
+
+			// Allow an optional human-readable descriptive name on the end of an Id.
+			// See http://msdn.microsoft.com/en-us/library/ms244717.aspx
+			var separatorIndex = info.Id.IndexOf (':');
+			if (separatorIndex != -1) {
+				info.Id = info.Id.Remove (separatorIndex);
+			}
+
+			if (attribute.HasProperties) {
+				foreach (var p in attribute.Properties) {
+					switch (p.Name) {
+					case "Scope":
+						info.Scope = p.Argument.Value as string;
+						break;
+					case "Target":
+						info.Target = p.Argument.Value as string;
+						break;
+					case "MessageId":
+						info.MessageId = p.Argument.Value as string;
+						break;
+					}
+				}
+			}
+
+			return true;
+		}
+
+		private class GlobalSuppressions
+		{
+			private readonly Dictionary<string, SuppressMessageInfo> _compilationWideSuppressions = new Dictionary<string, SuppressMessageInfo> ();
+
+			private readonly Dictionary<MetadataToken, Dictionary<string, SuppressMessageInfo>> _globalSuppressions =
```

Comment:
In the local suppressions you have ImmutableDictionary, here you have normal dictionary - any reason for the difference?

Reply:
I think that in Roslyn the `ImmutableDictionary` was used because of thread safety reasons, I've changed it to a `Dictionary`.